superdesk/users/async_service.py

```diff
@@ -0,0 +1,619 @@
+
+    async def on_created(self, docs: list[UsersResourceModel]) -> None:
+        """Send email to user with reset password token."""
+        await super().on_created(docs)
+        resetService = get_resource_service("reset_user_password")
+        activate_ttl = get_app_config("ACTIVATE_ACCOUNT_TOKEN_TIME_TO_LIVE")
+        for doc in docs:
+            user_dict = doc.to_dict()
+            if self.user_is_waiting_activation(doc) and user_dict.get("user_type") != "external":
+                user_id = user_dict.get("_id")
+                email = user_dict.get("email")
+                username = user_dict.get("username")
+                tokenDoc = {"user": user_id, "email": email}
+                id = resetService.store_reset_password_token(tokenDoc, email, activate_ttl, user_id)
+                if not id:
+                    raise SuperdeskApiError.internalError("Failed to send account activation email.")
+                tokenDoc.update({"username": username})
+
+                await send_activate_account_email(tokenDoc, activate_ttl)
+
+    async def on_update(self, updates: dict[str, Any], original: UsersResourceModel) -> None:
+        await super().on_update(updates, original)
+        if updates.get("first_name") or updates.get("last_name"):
+            user = original.to_dict()
+            updated_user = {
+                "first_name": user.get("first_name", ""),
+                "last_name": user.get("last_name", ""),
+                "username": user.get("username", ""),
+            }
+            if updates.get("first_name"):
+                updated_user["first_name"] = updates.get("first_name")
+            if updates.get("last_name"):
+                updated_user["last_name"] = updates.get("last_name")
+            updates["display_name"] = get_display_name(updated_user)
+
+    async def update_password(self, user_id: ObjectId | str, password: str):
+        """Update the user password.
+
+        Returns true if successful.
+        """
+        user = await self.find_by_id(user_id)
+
+        if not user:
+            raise SuperdeskApiError.unauthorizedError("User not found")
+
+        if not await self.is_user_active(user):
+            raise UserInactiveError()
+
+        updates = {
+            "password": get_hash(password, get_app_config("BCRYPT_GENSALT_WORK_FACTOR", 12)),
+            "password_changed_on": utcnow(),
+            LAST_UPDATED: utcnow(),
+        }
+
+        if self.user_is_waiting_activation(user):
+            updates["needs_activation"] = False
+
+        await self.update(user_id, updates=updates)
+
+    async def on_delete(self, doc: UsersResourceModel) -> None:
+        """
+        Overriding clean up reset password tokens:
+        """
+
+        await super().on_deleted(doc)
+        get_resource_service("reset_user_password").remove_all_tokens_for_email(doc.to_dict().get("email"))
+
+    async def _process_external_data(self, _data: dict[str, Any], update=False):
```

Comment:
These next 3 aren't required for this PR (they're needed for auth code, when auto-creating users from an external system). Can be left out of this new service for now